libbeat/processors/actions/add_tags_test.go

```diff
@@ -0,0 +1,108 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package actions
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/elastic/beats/libbeat/beat"
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/processors"
+)
+
+func TestAddTags(t *testing.T) {
+	multi := func(strs ...string) []string { return strs }
+	single := func(str string) []string { return multi(str) }
+
+	cases := map[string]struct {
+		event common.MapStr
+		want  common.MapStr
+		cfg   []string
+	}{
+		"create tags": {
+			event: common.MapStr{},
+			want:  common.MapStr{"tags": []string{"t1", "t2"}},
+			cfg:   single(`{tags: [t1, t2]}`),
+		},
+		"append to tags": {
+			event: common.MapStr{"tags": []string{"t1"}},
+			want:  common.MapStr{"tags": []string{"t1", "t2", "t3"}},
+			cfg:   single(`{tags: [t2, t3]}`),
+		},
+		"combine from 2 processors": {
+			event: common.MapStr{},
+			want:  common.MapStr{"tags": []string{"t1", "t2", "t3", "t4"}},
+			cfg: multi(
+				`{tags: [t1, t2]}`,
+				`{tags: [t3, t4]}`,
+			),
+		},
+		"with custom target": {
+			event: common.MapStr{},
+			want:  common.MapStr{"custom": []string{"t1", "t2"}},
+			cfg:   single(`{tags: [t1, t2], target: custom}`),
+		},
+		"different targets": {
+			event: common.MapStr{},
+			want:  common.MapStr{"tags1": []string{"t1"}, "tags2": []string{"t2"}},
+			cfg: multi(
+				`{target: tags1, tags: [t1]}`,
+				`{target: tags2, tags: [t2]}`,
+			),
+		},
+		"single tag config without array notation": {
+			event: common.MapStr{},
+			want:  common.MapStr{"tags": []string{"t1"}},
+			cfg:   single(`{tags: t1}`),
+		},
+	}
+
+	for name, test := range cases {
+		test := test
```

Comment:
This is a great example for me. Thanks! Just curious, is this `test := test` necessary here?

Reply:
It depends. I tend to put it here so to bind the name locally in case I enable `t.Parallel()`. functions/go-routines in a for-block accessing a loop variable will see the value being changed otherwise.

Reply:
handy tricks, I should use it more often.